routing/index.ts

```diff
@@ -0,0 +1,183 @@
+import {createAction} from '../actions';
+import {withTask} from '../tasks';
+import {HISTORY_PUSH_TASK} from '../history';
+
+type ComponentType = any;
+
+type RouteDefinition = {
+  [route: string]: ComponentType | [ComponentType, RouteDefinition]
+}
+
+/*
+ * Example usage:
+ *
+ * createRouter({
```

Comment:
What specifically do you not like about the API? I agree right now it's a bit opaque; the map from `String` to `Array | Component` is perhaps worse than a builder-type pattern.

However, one nice thing is that you can break up the declaration of this object into multiple files:

```
import {ChildComponent, CHILD_ROUTES}

const {stuff} = createRouter({
  '/child': [ChildComponent, CHILD_ROUTES]
});
```

The JSX version is a bad idea, in my opinion. What happens if you un-mount a configuration, for instance? It's very easy to adapt a function into a component, but much harder to go the other way. Agree it might be nice to shim, but beyond that I see no benefit.

I agree that this solution is not all-encompassing, but that's also kind of the point. We're just building functions up from a configuration and returning them in a convenient form. You can always wrap the functions before adding them to your reducer. It's better to provide common helpers that can be composed than to try to solve every use case.

We should consider exporting some of the helpers used in this file, so you could more easily write your own `createRouter` function.